diff --git a/src/com/android/contacts/detail/ContactLoaderFragment.java b/src/com/android/contacts/detail/ContactLoaderFragment.java
index e5decd7..65d4835 100644
--- a/src/com/android/contacts/detail/ContactLoaderFragment.java
+++ b/src/com/android/contacts/detail/ContactLoaderFragment.java
@@ -383,7 +383,8 @@ public class ContactLoaderFragment extends Fragment implements FragmentKeyListen
                     }
                 }
             } else if (SimContactsConstants.ACCOUNT_TYPE_PHONE.equals(accoutType)) {
-                copyToPhoneMenu.setVisible(false);
+		if(copyToPhoneMenu != null)
+                	copyToPhoneMenu.setVisible(false);
                 boolean hasPhoneOrEmail = hasPhoneOrEmailDate(mContactData);
                 if (MSimTelephonyManager.getDefault().isMultiSimEnabled()) {
                     if (hasPhoneOrEmail && hasEnabledIccCard(SimContactsConstants.SUB_1)) {
