diff --git a/init/init.c b/init/init.c
index bea1e76..39ba70d 100644
--- a/init/init.c
+++ b/init/init.c
@@ -85,6 +85,7 @@ static char hardware[32];
 static unsigned revision = 0;
 static char qemu[32];
 static char battchg_pause[32];
+static bool force_charging_mode = false;
 
 static struct action *cur_action = NULL;
 static struct command *cur_command = NULL;
@@ -828,6 +829,17 @@ static void import_kernel_nv(char *name, int for_emulator)
         cnt = snprintf(prop, sizeof(prop), "ro.boot.%s", boot_prop_name);
         if (cnt < PROP_NAME_MAX)
             property_set(prop, value);
+    /* DoPa (20140312) - the Hisense bootloader fails to use 'androidboot.mode=charger'
+     * to indicate that the tablet is booting solely to show the battery animations;
+     * this workaround looks for a particular value of 'power_reason' that corresponds
+     * to that state (Hisense appears to have done something similar in their version
+     * of 'init').
+     * Meticulus (20140928) - In order to standardize the devices I build for, switching
+     * to the lpm.rc method.
+     */
+    } else if (!strcmp(name, "power_reason") && !strcmp(value, "0x808800")) {
+        //property_set("ro.boot.mode", "charger");
+	force_charging_mode = true;
     }
 }
 
@@ -1064,6 +1076,7 @@ int audit_callback(void *data, security_class_t cls, char *buf, size_t len)
 
 static int charging_mode_booting(void)
 {
+    if(force_charging_mode) return true;
 #ifndef BOARD_CHARGING_MODE_BOOTING_LPM
     return 0;
 #else
